components/authz-service/storage/postgres/migration/sql/59_use_db_id_as_primary_key_in_iam_projects.up.sql

```diff
@@ -0,0 +1,26 @@
+BEGIN;
+
+ALTER TABLE iam_projects
+    DROP CONSTRAINT iam_projects_pkey; -- 'id' primary key
+ALTER TABLE iam_projects
+    ADD PRIMARY KEY (db_id); -- implies unique
+ALTER TABLE iam_projects
+    DROP CONSTRAINT iam_projects_db_id_key CASCADE; -- 'db_id' unique
+ALTER TABLE iam_projects
+    ADD UNIQUE (id);
+
+-- recreate all links
+-- Note: we've done this before; but in the separated steps, we've
+-- relied on the unique
+ALTER TABLE iam_project_rules
+    ADD CONSTRAINT iam_projects_db_id_fkey FOREIGN KEY (project_id) REFERENCES iam_projects(db_id);
+ALTER TABLE iam_staged_project_rules
+    ADD CONSTRAINT iam_staged_project_rules_project_id_fkey FOREIGN KEY (project_id) REFERENCES iam_projects(db_id);
+ALTER TABLE iam_role_projects
+    ADD CONSTRAINT iam_role_projects_project_id_fkey FOREIGN KEY (project_id) REFERENCES iam_projects(db_id);
+ALTER TABLE iam_policy_projects
+    ADD CONSTRAINT iam_policy_projects_project_id_fkey FOREIGN KEY (project_id) REFERENCES iam_projects(db_id);
+ALTER TABLE iam_statement_projects
+    ADD CONSTRAINT iam_statement_projects_project_id_fkey FOREIGN KEY (project_id) REFERENCES iam_projects(db_id);
```

Comment:
These aren't quite equlivant since they are missing `DEFERRABLE`.

Before migration:

```
    "iam_staged_project_rules_project_id_fkey" FOREIGN KEY (project_id) REFERENCES iam_projects(db_id) ON DELETE CASCADE DEFERRABLE
```

After migration:

```
    TABLE "iam_staged_rule_conditions" CONSTRAINT "iam_staged_rule_conditions_rule_db_id_fkey" FOREIGN KEY (rule_db_id) REFERENCES iam_staged_project_rules(db_id) ON DELETE CASCADE
```

However, after [reading up a bit](https://begriffs.com/posts/2017-08-27-deferrable-sql-constraints.html) it's not clear to me that this is a bad thing:

NOT DEFERRABLE | INITIALLY IMMEDIATE | INITIALLY DEFERRED
-- | -- | --
CHECK | row | row | row
NOT NULL | row | row | row
UNIQUE | row | statement | transaction
PRIMARY KEY | row | statement | transaction
FOREIGN KEY | statement | statement | transaction

This makes our transactions more flexible and I'm not sure for our use case we lose any safety.

Reply:
Actually, not sure which it defaults to, so not sure things are _actually_ different 🤔 They must be if the final output is different though.

Reply:
I need some education and/or a good link explaining the usage of `DEFERRABLE` and `INITIALLY DEFERRED`. From what I have seen so far just using `DEFERRABLE`, as we have done previously in our code, does _not_ mean we are making the constraint deferred, only that it could be deferred if some other statement (e.g. `SET CONSTRAINTS...`) asked for it.

Reply:
Thanks for paying attention to detail. I don't think we _need_ deferrable there; but I'd like to hear what @blakestier thinks about that.

`DEFERRABLE` means _"in a transaction, it's OK to be violated in the middle of it, as long as it's good on commit"_. 

This seems like a good in-depth intro: https://begriffs.com/posts/2017-08-27-deferrable-sql-constraints.html

Reply:
We can probably skip the deferrable, if we'd like. I was leaving it wherever it already was because we had other migrations that depended on the deferability, but given that those have already run by the time we hit this, it doesn't matter. Also, we can add deferability after the fact should we require it.

Reply:
mmm looks like we need them for upgrade on other tables, but not this one